src-browserify/api/sql.js

```diff
@@ -1,686 +1,697 @@
-  loc = loc.slice(0, loc.length - 1);
-  if(loc == 'file') {
-    loc = 'https';
-  }
-
-  this.ajax = options.ajax || (typeof(jQuery) !== 'undefined' ? jQuery.ajax: reqwest);
-  if(!this.ajax) {
-    throw new Error("jQuery or reqwest should be loaded");
-  }
-
-  this.options = _.defaults(options, {
-    version: 'v2',
-    protocol: loc,
-    jsonp: typeof(jQuery) !== 'undefined' ? !jQuery.support.cors: false
-  })
-
-  if (!this.options.sql_api_template) {
-    var opts = this.options;
-    var template = null;
-    if(opts && opts.completeDomain) {
-      template = opts.completeDomain;
-    } else {
-      var host = opts.host || 'cartodb.com';
-      var protocol = opts.protocol || 'https';
-      template = protocol + '://{user}.' + host;
+// NOTE this does not return a SQL directly, but a wrapper, to inject the dependencies
+// e.g. var SQL = require('./sql')({ jQuery: $ /* or reqwest: reqwest */ });
+// @param {Object} _Promise
+// @param {Object} opts hash with keys:
+//   jQuery: {Object} (Optional) has precedence over reqwest if both are present
+//   reqwest: {Object} (Optional)
+module.exports = function(_Promise, opts) {
+  if (!_Promise) throw new Error('_Promise is required');
+  if (!opts.jQuery && !opts.reqwest) throw new Error('opts.jQuery or opts.reqwest is required');
+  var jQuery = opts.jQuery;
+  var reqwest = opts.reqwest;
+
+  function SQL(options) {
+    if(cartodb === this || window === this) {
+      return new SQL(options);
+    }
+    if(!options.user) {
+      throw new Error("user should be provided");
+    }
+    var loc = new String(window.location.protocol);
+    loc = loc.slice(0, loc.length - 1);
+    if(loc == 'file') {
+      loc = 'https';
     }
-    this.options.sql_api_template = template;
-  }
-}
-
-SQL.prototype._host = function() {
-  var opts = this.options;
-  return opts.sql_api_template.replace('{user}', opts.user) + '/api/' +  opts.version + '/sql';
-},
-
-/**
- * var sql = new SQL('cartodb_username');
- * sql.execute("select * from {{ table }} where id = {{ id }}", {
- *    table: 'test',
- *    id: '1'
- * })
- */
-SQL.prototype.execute = function(sql, vars, options, callback) {
-
-  //Variable that defines if a query should be using get method or post method
-  var MAX_LENGTH_GET_QUERY = 1024;
-
-  var promise = new _Promise();
-  if(!sql) {
-    throw new TypeError("sql should not be null");
-  }
-  // setup arguments
-  var args = arguments,
-  fn = args[args.length -1];
-  if(_.isFunction(fn)) {
-    callback = fn;
-  }
-  options = _.defaults(options || {}, this.options);
-  var params = {
-    type: 'get',
-    dataType: 'json',
-    crossDomain: true
-  };
-
-  if(options.cache !== undefined) {
-    params.cache = options.cache;
-  }
 
-  if(options.jsonp) {
-    delete params.crossDomain;
-    if (options.jsonpCallback) {
-      params.jsonpCallback = options.jsonpCallback;
+    this.ajax = options.ajax || (typeof(jQuery) !== 'undefined' ? jQuery.ajax: reqwest);
```

Comment:
I know we just moved this, but it'd probably need to be `reqwest.compat`. (We're doing it [here](https://github.com/CartoDB/cartodb.js/blob/60285edb65f6575c8a34799fde55308751251c5f/src/geo/layer_definition.js#L68))